controllers/core/openstackversion_controller.go

```diff
@@ -0,0 +1,295 @@
+		if !compareStringPointers(controlPlane.Status.ContainerImages.OvnControllerImage, instance.Status.ContainerImages.OvnControllerImage) ||
+			!controlPlane.Status.Conditions.IsTrue(corev1beta1.OpenStackControlPlaneOVNReadyCondition) {
+			instance.Status.Conditions.Set(condition.FalseCondition(
+				corev1beta1.OpenStackVersionMinorUpdateOVNControlplane,
+				condition.RequestedReason,
+				condition.SeverityInfo,
+				corev1beta1.OpenStackVersionMinorUpdateReadyRunningMessage))
+			Log.Info("Minor update for OVN Controlplane in progress")
+			return ctrl.Result{}, nil
+		}
+		instance.Status.Conditions.MarkTrue(
+			corev1beta1.OpenStackVersionMinorUpdateOVNControlplane,
+			corev1beta1.OpenStackVersionMinorUpdateReadyMessage)
+
+		// minor update for Controlplane in progress
+		// we only check keystone here as it will only get updated during this phase
+		// FIXME: add checks to all images on the Controlplane here once conditions and observedGeneration work are finished
+		if !compareStringPointers(controlPlane.Status.ContainerImages.KeystoneAPIImage, instance.Status.ContainerImages.KeystoneAPIImage) ||
+			!controlPlane.IsReady() {
+			instance.Status.Conditions.Set(condition.FalseCondition(
+				corev1beta1.OpenStackVersionMinorUpdateControlplane,
+				condition.RequestedReason,
+				condition.SeverityInfo,
+				corev1beta1.OpenStackVersionMinorUpdateReadyRunningMessage))
+			Log.Info("Minor update for Controlplane in progress")
+			return ctrl.Result{}, nil
+		}
+		// TODO minor update for Dataplane in progress goes here
+
+		instance.Status.Conditions.MarkTrue(
+			corev1beta1.OpenStackVersionMinorUpdateControlplane,
+			corev1beta1.OpenStackVersionMinorUpdateReadyMessage)
+	}
+
+	if controlPlane.IsReady() {
+		Log.Info("Setting DeployedVersion")
+		instance.Status.DeployedVersion = &instance.Spec.TargetVersion
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *OpenStackVersionReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	versionFunc := handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, o client.Object) []reconcile.Request {
+		Log := r.GetLogger(ctx)
+		versionList := &corev1beta1.OpenStackVersionList{}
+
+		result := []reconcile.Request{}
+
+		listOpts := []client.ListOption{
+			client.InNamespace(o.GetNamespace()),
+		}
+		if err := r.Client.List(ctx, versionList, listOpts...); err != nil {
+			Log.Error(err, "Unable to retrieve OpenStackVersion")
+			return nil
+		}
+
+		for _, i := range versionList.Items {
+			name := client.ObjectKey{
+				Namespace: o.GetNamespace(),
+				Name:      i.Name,
+			}
+			result = append(result, reconcile.Request{NamespacedName: name})
+		}
+		if len(result) > 0 {
+			Log.Info("Reconcile request for:", "result", result)
+			return result
+		}
+		return nil
+	})
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Watches(&corev1beta1.OpenStackControlPlane{}, versionFunc).
```

Comment:
Could you explain the logic behind this? How we are watching the the OpenStackControlPlane objects by querying the OpenStackVersion objects in the versionFunc?